test/rules/no-unnecessary-initializer/test.ts.lint

```diff
@@ -0,0 +1,24 @@
+let a: string | undefined = undefined;
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [0]
+var b: string | undefined = undefined;
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [0]
+
+for (let c: number | undefined = undefined; c < 2; c++) {}
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [0]
+
+let d;
+
+const e = undefined;
+
+function f(x: string | undefined = undefined, bar: number) {}
```

Comment:
On second thought, on this function the `= undefined` should lead to failure [0].
If parameter `bar` is required, you need to pass a value for `x`. So is makes no sense to set a default of `undefined`, because the default value is only used when the passed value for that parameter is `undefined`.